tfhe/src/integer/server_key/radix_parallel/shift.rs

```diff
@@ -68,35 +89,105 @@ impl ServerKey {
+        }
+
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
+        }
+
+        let message_modulus = self.key.message_modulus.0 as u64;
+
+        // Since we require that carries are empty,
+        // we can use the bivariate bps to shift and propagate in parallel at the same time
+        // instead of first shifting then propagating
+        //
+        // The first block is done separately as it does not
+        // need to recuperate the shifted bits from its next block,
+        // and also that way is does not need a special case for when rotations == 0
+        let create_blocks_using_bivariate_pbs = || {
+            let lut =
+                self.key
+                    .generate_accumulator_bivariate(|mut current_block, mut next_block| {
+                        // left shift so as not to lose
+                        // bits when shifting right afterwards
+                        current_block <<= num_bits_in_block;
+                        next_block <<= num_bits_in_block;
+
+                        current_block >>= shift_within_block;
+                        next_block >>= shift_within_block;
+
+                        // The way of gettint caryy / message is reversed compared
+                        // to the usual way but its normal:
+                        // The message is in the upper bits, the carry in lower bits
+                        let message_of_current_block = current_block >> num_bits_in_block;
+                        let carry_of_previous_block = next_block % message_modulus;
+
+                        message_of_current_block + carry_of_previous_block
+                    });
+            let partial_blocks = ct.blocks[..num_blocks - rotations]
+                .par_windows(2)
+                .map(|blocks| {
+                    // We are right-shifting,
+                    // so we get the bits from the next block in the vec
+                    let (current_block, next_block) = (&blocks[0], &blocks[1]);
+                    self.key
+                        .unchecked_apply_lookup_table_bivariate(current_block, next_block, &lut)
+                })
+                .collect::<Vec<_>>();
+            partial_blocks
+        };
+
+        let shift_last_block = || {
+            let mut block = ct.blocks[num_blocks - rotations - 1].clone();
+            self.key
+                .unchecked_scalar_right_shift_assign(&mut block, shift_within_block as u8);
+            self.key.message_extract_assign(&mut block);
+            block
+        };
+        let (partial_blocks, block) =
+            rayon::join(create_blocks_using_bivariate_pbs, shift_last_block);
+
+        debug_assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
+        for (block, shifted_block) in ct.blocks[..num_blocks - rotations - 1]
+            .iter_mut()
+            .zip(partial_blocks.into_iter())
+        {
+            *block = shifted_block
         }
```

Comment:
could be parallelized ? not sure we win a lot

Reply:
Entering a thread/switching context to a thread just to replace (replace + implicit drop) a value won't provide any gains